### PR TITLE
Add EW actions, fix action ID logic

### PR DESCRIPTION
**Pull request by MoFroGT**
Only known issue is positional checks for reaper are not working.

### CODE REVIEW

**Comment by MoFroGT** [comment]
#32 would be addressed by this.

**Comment by notdryft** [comment]
> Only known issue is positional checks for reaper are not working.

It may be because you put the reaper if check inside the if block for samurai!

**Comment by MoFroGT** [comment]
You're not wrong... however SAM/RPR positionals are pretty broken regardless. The old version used kenki generation to determine if a SAM positional was hit or not, but this is different in EW (positional only changes potency). I'm unable to determine what, if anything, designates a positional hit or miss in the incoming ACT data.

**Comment by notdryft** [comment]
So for SAM it looks like it now shares the same succeed code as ninja, which is '11B', so the following code works:

```javascript
  if (samuraiPositionals.includes(action.actionID)) {
    // samurai rear/flank check
    resources.positionalActionCount++

    const succeedCode = '11B'
    return logParameter.slice(8, 22).includes(succeedCode)
  }
```

I don't have reaper unlocked so can't check... i'll continue to investigate on my fork anyway.

**Comment by notdryft** [comment]
Actually it's more complicated, this code doesn't appear if under meikyo...

**Comment by MoFroGT** [comment]
Would explain why I didn't see it - I did my testing with Meikyo.

**Comment by notdryft** [comment]
There might be some way around using another log code, because with 22 (network ability), here is what I got 👍 

image

Tried a mix of "ok" (positional hit) and "ko" (positional missed) raw, under meikyo, true north (tn) and both. It turns out there is no way to tell if the pos is hit or not as long as meikyo is on... So '11B' can still be used, but not all the time...

Searched through https://github.com/quisquous/cactbot/blob/main/docs/LogGuide.md to see what we could use instead. Gauge is useless as it only contains the number of sens and how much the gauge is filled. I'm afraid the potency of positional check is baked in the damage calculation and can't be found anywhere now :/

**Comment by MoFroGT** [comment]
What about the 2nd position in the flags? Looks like it's 48710003 whenever positional is registered as a hit and 44710003 when it is not?

**Comment by notdryft** [comment]
OMG, I'm blind, I didn't see it! This works like a charm:

```javascript
  if (samuraiPositionals.includes(action.actionID)) {
    // samurai rear/flank check
    resources.positionalActionCount++

    return logParameter[6].slice(0, 2) === '48'; // flags
  }
```

Just tested it in Bozja, the flag returned when the monster has no positional is also 48. It looks like it's bulletproof.

**Comment by notdryft** [comment]
Right above this code is trick attack in fact, which test against '1E71', so it could be for Samurai, to match the style:

```javascript
if (samuraiPositionals.includes(action.actionID)) {
    // samurai rear/flank check
    resources.positionalActionCount++

    const succeedCode = '4871' // 48710003 48710103 48710203
    // failedCode: 44710103 44710203
    return logParameter[6].includes(succeedCode)
  }
```

**Comment by MoFroGT** [comment]
Excellent! I'll see if I can't uncover the same for reaper.

**Comment by MoFroGT** [comment]
There are, in fact, two different success codes for RPR. 

`  if (reaperPositionals.includes(action.actionID)) {
    // reaper rear/flank check
    resources.positionalActionCount++

    const succeedCode1 = 'B710' // B710003 B710103 
    const succeedCode2 = 'D710' // D710003 D710103
    // failedCode: 710003 710003
    return logParameter[6].includes(succeedCode1) + logParameter[6].includes(succeedCode2)
  }`

The one that starts with B indicates correct use of the Enhanced Gibbet or Enhanced Gallows buff, while D indicates an un-buffed positional hit. The first Gibbet/Gallows executed in any encounter will always be un-buffed, so it would be incorrect to assume all un-buffed hits are errors. However, beyond the first one, any misuse of the buff should be classified as an error.

Do you know if there is a way to check whether the buff is active or not in the logic? If not, then this may be the best we can do for now.

**Comment by notdryft** [comment]
You could use code 26 (Buff) and 30 (BuffRemove) in `parseLogLine` (src/parseMessage.js) (https://github.com/quisquous/cactbot/blob/main/docs/LogGuide.md#line-26-0x1a-networkbuff) and add a flag to `resources.reaper` maybe.

File: src/handleAction.js [hunk 1]
```diff
@@ -207,15 +207,19 @@ const checkPositional = (action, logParameter) => {
...
+    return logParameter[6].includes(succeedCode)
+  }
+  if (reaperPositionals.includes(action.actionID)) {
+    // reaper rear/flank check
+    resources.positionalActionCount++
+
+    const succeedCode1 = 'B710' // B710003 B710103
+    const succeedCode2 = 'D710' // D710003 D710103
+    // failedCode: 710003 710003
+    return logParameter[6].includes(succeedCode1) + logParameter[6].includes(succeedCode2)
```
Comment (notdryft) [inline src/handleAction.js:222]: Don't you mean `||` there instead of `+` ?

Comment (MoFroGT) [reply src/handleAction.js:222]: JS dynamic casting makes them the same in this context I think

Comment (MoFroGT) [reply src/handleAction.js]: Fixed it for the sake of accuracy

**Comment by sgosiaco** [comment]
I just want to mention that I've found the most reliable way to do the SAM positional check is to check for three codes.
The `11B` code as other melees
The `4871` code for when under meikyo
And the `2171` code for when it's not part of a combo

In EW, both Gekko and Kasha got updated to have a positional even when not under combo.
So something like this
```
 if (samuraiPositionals.includes(action.actionID)) {
    // samurai rear/flank check
    resources.positionalActionCount++
    const succeedCode = '11B'
    const meikyoCode = '4871'
    const noComboCode = '2171'
    return logParameter.slice(8, 22).includes(succeedCode) || logParameter[6].includes(meikyoCode) || logParameter[6].includes(noComboCode)
  }
```

**Comment by ramram1048** [comment]
@MoFroGT @notdryft @sgosiaco Can i”we merge this? (I didn’t check the code or test anyway)

**Comment by notdryft** [comment]
Yes it can!

**Pull request merged**